fhevm/operators_rand.go

```diff
@@ -84,58 +82,15 @@ func generateRandom(environment EVMEnvironment, caller common.Address, resultTyp
 	if err != nil {
 		return nil, err
 	}
-	// The RNG nonce bytes are of size chacha20.NonceSizeX, which is assumed to be 24 bytes (see init() above).
-	// Since uint256.Int.z[0] is the least significant byte and since uint256.Int.Bytes32() serializes
-	// in order of z[3], z[2], z[1], z[0], we want to essentially ignore the first byte, i.e. z[3], because
-	// it will always be 0 as the nonce size is 24.
-	cipher, err := chacha20.NewUnauthenticatedCipher(seed.Bytes(), currentRngNonceBytes[32-chacha20.NonceSizeX:32])
-	if err != nil {
-		return nil, err
-	}
 
-	// XOR a byte array of 0s with the stream from the cipher and receive the result in the same array.
-	// Apply upperBound, if set.
-	var randUint uint64
-	switch resultType {
-	case tfhe.FheUint4:
-		randBytes := make([]byte, 1)
-		cipher.XORKeyStream(randBytes, randBytes)
-		randUint = uint64(randBytes[0])
-		randUint = uint64(applyUpperBound(randUint, 4, upperBound))
-	case tfhe.FheUint8:
-		randBytes := make([]byte, 1)
-		cipher.XORKeyStream(randBytes, randBytes)
-		randUint = uint64(randBytes[0])
-		randUint = uint64(applyUpperBound(randUint, 8, upperBound))
-	case tfhe.FheUint16:
-		randBytes := make([]byte, 2)
-		cipher.XORKeyStream(randBytes, randBytes)
-		randUint = uint64(binary.BigEndian.Uint16(randBytes))
-		randUint = uint64(applyUpperBound(randUint, 16, upperBound))
-	case tfhe.FheUint32:
-		randBytes := make([]byte, 4)
-		cipher.XORKeyStream(randBytes, randBytes)
-		randUint = uint64(binary.BigEndian.Uint32(randBytes))
-		randUint = uint64(applyUpperBound(randUint, 32, upperBound))
-	case tfhe.FheUint64:
-		randBytes := make([]byte, 8)
-		cipher.XORKeyStream(randBytes, randBytes)
-		randUint = uint64(binary.BigEndian.Uint64(randBytes))
-		randUint = uint64(applyUpperBound(randUint, 64, upperBound))
-	default:
-		return nil, fmt.Errorf("generateRandom() invalid type requested: %d", resultType)
-	}
-
-	// Trivially encrypt the random integer.
-	randCt := new(tfhe.TfheCiphertext)
-	randBigInt := big.NewInt(0)
-	randBigInt.SetUint64(randUint)
-	randCt.TrivialEncrypt(*randBigInt, resultType)
-	insertCiphertextToMemory(environment, randCt)
+	randCt, err := tfhe.GenerateObliviousPseudoRandom(resultType, *(*uint64)(unsafe.Pointer(&seed.Bytes()[0])), numberOfBits)
```

Comment:
Maybe not critical for this PR, but AFAIU, the seed can just be a counter. Right now, we hash a fixed global seed and the caller address, leading to the same seed every time. We should double check with tfhe-rs and, if that's the case, change the seed.

Reply:
OK you mean we could use directly `globalRngSeed` ?

Reply:
Is `globalRngSeed` incremented after each use?

Reply:
Maybe `nextRngNonce` is a better choice. I don't see where we use it right now

Reply:
I am getting it right, that `globalRngSeed`/`nextRngNonce` is local state? For the coprocessor we might want to avoid this, and instead use a counter from Solidity. This would make the coprocessor computation stateless and reproducible. To generate the symbolic values for random encryptions, we need a Solidity counter (plaintext integer) anyways. This counter could be used here as well.